rts/Sim/MoveTypes/Systems/GroundMoveSystem.cpp

```diff
@@ -66,33 +139,50 @@ void GroundMoveSystem::Update() {
     }
 	{
         SCOPED_TIMER("Sim::Unit::MoveType::4::ProcessCollisionEvents");
-        auto view = Sim::registry.view<GroundMoveType>();
-        view.each([](GroundMoveType& unitId){
-            CUnit* unit = unitHandler.GetUnit(unitId.value);
-            AMoveType* moveType = unit->moveType;
-            moveType->ProcessCollisionEvents();
+
+        process_thread_data<UnitCrushEvent>(comp.killUnits, [](UnitCrushEvent& event){
+            event.collidee->Kill(event.collider, event.crushImpulse, true);
+        });
+        process_thread_data<FeatureCrushEvent>(comp.killFeatures, [](FeatureCrushEvent& event) {
+            event.collidee->Kill(event.collider, event.crushImpulse, true);
+        });
+        process_thread_data<UnitCollisionEvent>(comp.collidedUnits, [](UnitCollisionEvent& event) {
+            eventHandler.UnitUnitCollision(event.collider, event.collidee);
+        });
+        process_thread_data<FeatureCollisionEvent>(comp.collidedFeatures, [](FeatureCollisionEvent& event) {
+            eventHandler.UnitFeatureCollision(event.collider, event.collidee);
+        });
+        process_thread_data<FeatureMoveEvent>(comp.moveFeatures, [](FeatureMoveEvent& event) {
+            quadField.RemoveFeature(event.collidee);
+            event.collidee->Move(event.moveImpulse, true);
+            quadField.AddFeature(event.collidee);
         });
 	}
 	{
-        SCOPED_TIMER("Sim::Unit::MoveType::5::UpdateST");
+        SCOPED_TIMER("Sim::Unit::MoveType::5::Update");
         auto view = Sim::registry.view<GroundMoveType>();
-        view.each([](GroundMoveType& unitId){
+        // view.each([](GroundMoveType& unitId){
+        for_mt(0, view.size(), [&view, &comp](const int i){
+            auto curThread = ThreadPool::GetThreadNum();
+
+            auto entity = view.storage<GroundMoveType>()[i];
+            auto unitId = view.get<GroundMoveType>(entity);
+
             CUnit* unit = unitHandler.GetUnit(unitId.value);
             AMoveType* moveType = unit->moveType;
-
             if (moveType->Update())
-                eventHandler.UnitMoved(unit);
-
-            // this unit is not coming back, kill it now without any death
-            // sequence (s.t. deathScriptFinished becomes true immediately)
-            if (!unit->pos.IsInBounds() && (unit->speed.w > MAX_UNIT_SPEED))
-                unit->ForcedKillUnit(nullptr, false, true);
+                comp.movedUnits[curThread].emplace_back(UnitMovedEvent(unitId.value));
 
             #ifndef NDEBUG
             unit->SanityCheck();
             #endif
         });
-	}
+
+        process_thread_data<UnitMovedEvent>(comp.movedUnits, [](UnitMovedEvent& event){
+            CUnit* unit = unitHandler.GetUnit(event.id);
```

Comment:
no null check needed? also perhaps consider `const CUnit* unit`, and maybe `const UnitMovedEvent& event`

Reply:
I fixed this when switching to using the job id.